### PR TITLE
THRIFT-4759:add thrift export to fix windows shared lib building

### PR DESCRIPTION
**Pull request by cyyever**
add thrift export to fix windows shared lib building

### CODE REVIEW

**Comment by jeking3** [comment]
We're moving towards cmake but it's going to take a while to get everything that autoconf does for us working in cmake... in the meantime thanks for both.  The "make dist" job in Travis CI failed and it looks related.

File: lib/cpp/CMakeLists.txt [hunk 1]
```diff
@@ -23,6 +23,10 @@ REQUIRE_BOOST_HEADERS()
 
 include_directories(src)
 
+if(NOT BUILD_SHARED_LIBS)
+  add_definitions("-DTHRIFT_STATIC_DEFINE")
```
Comment (jeking3) [inline lib/cpp/CMakeLists.txt]: Seems oddly named... suggestion...
```
if (BUILD_SHARED_LIBS)
    add_definitions("-DTHRIFT_LINK_SHARED")
endif ()

**Comment by cyyever** [comment]
> 
> 
> We're moving towards cmake but it's going to take a while to get everything that autoconf does for us working in cmake... in the meantime thanks for both. The "make dist" job in Travis CI failed and it looks related.

I forgot to change autotool scripts,I will re-push it.

File: lib/cpp/CMakeLists.txt (hunk 1 above)
Comment (cyyever) [reply lib/cpp/CMakeLists.txt]: THRIFT_STATIC_DEFINE is generated by cmake.In order to switch to cmake smoothly in the future,I don't want to change it.

**Comment by cyyever** [comment]
@jeking3 by the way,are there any plans to merge fbthrift code?

**Comment by jeking3** [comment]
Boost leads the way in terms of correctness here, just a suggestion for long term.

**Comment by jeking3** [comment]
I think I prefer the other PR for this issue.  We don't use autoconf for any build that needs dllimport/export - we only use cmake.  Therefore a cmake-centric solution is appropriate, especially if nobody wants to tackle symbol visibility on non-windows systems right now.

**Comment by cyyever** [comment]
> 
> 
> We're moving towards cmake but it's going to take a while to get everything that autoconf does for us working in cmake... in the meantime thanks for both. The "make dist" job in Travis CI failed and it looks related.
AppVeyor passed,Travis CI failed for networking problems.

**Comment by jeking3** [comment]
fbthrift code changes would need to be submitted by the fbthrift team.  We cannot pull them in.

**Comment by jeking3** [comment]
Also rebase on master which is more stable... I fixed the concurrency_test.

**Comment by emmenlau** [comment]
Looking forward, thanks!

**Comment by cyyever** [comment]
> 
> 
> Also rebase on master which is more stable... I fixed the concurrency_test.

just rebase on master.I am not familiar with appveyor scripts.

**Comment by jeking3** [comment]
In the file appveyor.yml there is a directive that sets the MSVC 2017 build to static.  Change it.  None of the windows MSVC build tested your changes because both of them disable shared library builds.

**Comment by cyyever** [comment]
> 
> 
> In the file appveyor.yml there is a directive that sets the MSVC 2017 build to static. Change it. None of the windows MSVC build tested your changes because both of them disable shared library builds.

fixed

**Comment by ericu65** [comment]
This PR throws the following warning in MSVC:
warning LNK4049: locally defined symbol ?GlobalOutput@thrift@apache@@3VTOutput@12@A (class apache::thrift::TOutput apache::thrift::GlobalOutput) imported

**Comment by ericu65** [comment]
@cyyever @jeking3 note above is in regard to static libraries on MSVC 15.

**Comment by goorjn** [comment]
@cyyever @jeking3 wouldn't it have been more convenient to move the TSSLSocketFactory::setManualOpenSSLInitialization's body to cpp rather than explicitly exporting the variable ? (see https://stackoverflow.com/questions/2479784/exporting-static-data-in-a-dll)
Maybe there is a similar trick to avoid to have to explicitly export GlobalOutput variable, but I still haven't looked for it.
By the way, I would have preferred to have default static linking for windows, just like previous versions, rather than to have to #define THRIFT_STATIC_DEFINE in my projects to do so.

**Comment by Jens-G** [comment]
https://issues.apache.org/jira/browse/LEGAL-538

**Pull request merged**